In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Problem 4: Document Clustering

Document clustering is a vital application of clustering algorithms. It can be utilized to automatically categorize search results, facilitating the discovery of relevant information.

In this problem, you will cluster articles published on [Daily Kos](https://www.dailykos.com/), an American political blog.

The file dailykos.csv includes data on 3,430 news articles or blogs posted on Daily Kos in 2004, leading up to the United States Presidential Election.

In [2]:
# load the data
url = 'https://raw.githubusercontent.com/um-perez-alvaro/Data-Science-Theory/master/Data/dailykos.csv'
data = pd.read_csv(url)
data

abandon  abc  ability  abortion  absolute  abstain  abu  abuse  accept  \
0           0    0        0         0         0        0    0      0       0   
1           0    0        0         0         0        0    0      0       0   
2           0    0        0         0         0        1    0      0       0   
3           0    0        0         0         0        0    0      0       0   
4           0    0        0         0         0        0    0      0       0   
...       ...  ...      ...       ...       ...      ...  ...    ...     ...   
3425        0    0        0         0         0        0    0      0       0   
3426        0    0        0         0         0        0    0      0       0   
3427        0    0        0         0         0        0    0      0       0   
3428        0    0        0         0         0        0    0      0       0   
3429        0    0        0         0         0        0    0      0       0   

      access  ...  yeah  year  yesterday  york  youll  young  youre  youve  \
0          0  ...     0     0          0     0      0      0      0      0   
1          0  ...     0     0          0     0      0      0      0      0   
2          0  ...     0     0          0     0      0      0      0      0   
3          0  ...     0     0          2     0      0      1      0      0   
4          0  ...     0     1          1     0      0      1      0      0   
...      ...  ...   ...   ...        ...   ...    ...    ...    ...    ...   
3425       0  ...     0     0          0     0      0      0      0      0   
3426       0  ...     0     0          0     0      0      0      0      0   
3427       0  ...     0     0          0     1      0      0      0      0   
3428       0  ...     0     0          0     0      0      0      0      0   
3429       0  ...     0     0          0     0      0      0      0      0   

      zogby  zone  
0         0     1  
1         0     0  
2         0     0  
3         0     0  
4         0     0  
...     ...   ...  
3425      0     0  
3426      0     0  
3427      0     0  
3428      0     0  
3429      0     0  

[3430 rows x 1545 columns]

Each feature in the dataset represents a word that has appeared in at least 50 distinct articles (1,545 words in total). The feature values for each document correspond to the frequency with which the respective word has appeared in the document.

In [ ]:
list_of_words = np.array(data.columns)
list(list_of_words)

## Part 1: k-means clustering

Cluster the articles using k-means and ensure that you select an appropriate number of clusters, for instance by utilizing techniques like the 'elbow' method.

In [17]:
X = data.to_numpy().astype('float')

## Part 2

Find the 10 most common words in each cluster.

In 2004, Howard Dean was a candidate for the Democratic nomination for the President of the United States, while John Kerry won the Democratic nomination with John Edwards as his running mate (the Vice President nominee). Considering this information, which cluster most accurately corresponds to the Democratic Party?

Which k-means cluster is most closely associated with the Iraq War?

## Part 3: Hierarchical Clustering

Now, apply hierarchical clustering (this process may take a while; why?). Choose the number of clusters to be equal to the number of clusters utilized in Parts 1 and 2.